### ISSUE TITLE
Antialiasing tests fail with vtk >=9.0.0

### ISSUE THREAD

**Issue by jasonb5** [CDAT/vcs#443]
Antiasliasing unittest is failing (https://circleci.com/gh/CDAT/vcs/3293).

https://github.com/CDAT/vcs/blob/master/tests/test_vcs_antialiasing.py

```
b'======================================================================'
b'FAIL: testAntiAliasing (test_vcs_antialiasing.TestVCSAntialias)'
b'----------------------------------------------------------------------'
b'Traceback (most recent call last):'
b'  File "/home/circleci/project/tests/test_vcs_antialiasing.py", line 11, in testAntiAliasing'
b'    self.assertEqual(x.getantialiasing(), 8)'
b'AssertionError: 0 != 8'
b'-------------------- >> begin captured logging << --------------------'
b'asyncio: DEBUG: Using selector: EpollSelector'
b'--------------------- >> end captured logging << ---------------------'
b'----------------------------------------------------------------------'
b'Ran 1 test in 0.012s'
b'FAILED (failures=1)'
FAILED...cmd: ['nosetests', '-A', 'not vtk_ui', '-s', '/home/circleci/project/tests/test_vcs_antialiasing.py']
```

**Comment by jasonb5** [comment]
I created two environments with the following:

Mesalib 17.3.9
```bash
conda create -n vcs-aa-mesa-17.3.9 -c conda-forge/label/vtk-dev -c cdat/label/vtk-dev -c cdat/label/nightly -c conda-forge "python>=3.7,<3.8" vcs "mesalib=17.3.9" pytest pillow
```

Mesalib 18.3.1
```bash
conda create -n vcs-aa-mesa-18.3.1 -c conda-forge/label/vtk-dev -c cdat/label/vtk-dev -c cdat/label/nightly -c conda-forge "python>=3.7,<3.8" vcs "mesalib=18.3.1" pytest pillow
```
Next I ran this script to generate the following images; baseline (1), aa=0 (2), aa=8 (3), difference of (1, 2) and (1, 3). I would expect the (1, 3) to appear black since there should be no difference and (1, 2) not show black since there should be a difference.

`python vcs-test.py 17.3.9` and `python vcs-test.py 18.3.1`

```python
import vcs
import cdms2
import argparse
from PIL import Image, ImageChops

parser = argparse.ArgumentParser()

parser.add_argument('version', type=str)

args = parser.parse_args()

with cdms2.open('clt.nc') as f:
    data = f('clt')

def plot(version, aa):
    extra = ''
    x = vcs.init(bg=True)

    if aa is not None:
        x.setantialiasing(aa)

    x.plot(data,bg=True)

    if aa is None:
        aa = x.backend.renWin.GetMultiSamples()

        extra = '-baseline'

    fname = f'vcs-aa-mesa_{version}-{aa}{extra}.png'

    x.png(fname)

    x.close()

    return fname

baseline = plot(args.version, None)
aa0 = plot(args.version, 0)
aa1 = plot(args.version, 8)

img_baseline = Image.open(baseline)
img_aa0 = Image.open(aa0)
img_aa1 = Image.open(aa1)

diff0 = ImageChops.difference(img_baseline, img_aa0)
diff0.save(f'vcs-aa-mesa-diff-aa0.png')

diff1 = ImageChops.difference(img_baseline, img_aa1)
diff1.save(f'vcs-aa-mesa-diff-aa8.png')
```
The results on my local linux box are as expected. Next step is to spin the circleci test with ssh and run the script there.

**Comment by jasonb5** [comment]
Created environment using:

```bash
conda create -n vcs-mesa-8.3.1 -c cdat/label/vtk_dev -c conda-forge/label/vtk_dev -c conda-forge vcs=8.2.2020.04.20.14.05.g2383318 "vtk=9.0.0.rc1" "mesalib=18.3.1" pillow
```
Conda list `conda list "vcs|vtk|mesalib"` should produce the following:

```bash
# packages in environment at /home/titters/conda/envs/vcs-test:
#
# Name                    Version                   Build  Channel
mesalib                   18.3.1               h590aaf7_0    conda-forge
vcs                       8.2.2020.04.20.14.05.g2383318    pyh9f0ad1d_0    cdat/label/vtk_dev
vtk                       9.0.0.rc1       with_osmesa_py38he42c7bc_4  [mesalib]  conda-forge/label/vtk_dev

```

**Comment by jasonb5** [comment]
I ran the following script that generates a baseline image and then attempts to set antialiasing to 8.

Source file: [clt.nc](https://cdat.llnl.gov/cdat/sample_data/clt.nc)

`python vcs-test.py 18.3.1`

```python
import vcs
import cdms2
import argparse
from PIL import Image, ImageChops

parser = argparse.ArgumentParser()

parser.add_argument('version', type=str)

args = parser.parse_args()

with cdms2.open('clt.nc') as f:
    data = f('clt')

def plot(version, aa=None, gm=None):
    x = vcs.init(bg=True)

    if aa is not None:
        x.setantialiasing(aa)
    else:
        aa = f'baseline'

    x.plot(data, gm, bg=True)

    post_aa = x.backend.renWin.GetMultiSamples()

    fname = f'vcs-aa-mesa_{version}-{aa}-{post_aa}.png'

    x.png(fname)

    x.close()

    return fname

baseline = plot(args.version)
aa8 = plot(args.version, 8)

img1 = Image.open(baseline)
img2 = Image.open(aa8)

diff1 = ImageChops.difference(img1, img2)
diff1.save('vcs-aa-mesa-baseline-aa8.png')
```
Generated the following plots:
![vcs-aa-mesa_18 3 1-8-0](https://user-images.githubusercontent.com/2191036/80265902-11a7c480-864e-11ea-99c7-56398d6623d0.png)
![vcs-aa-mesa_18 3 1-baseline-0](https://user-images.githubusercontent.com/2191036/80265907-140a1e80-864e-11ea-91bc-2fb86f36810c.png)
![vcs-aa-mesa-baseline-aa8](https://user-images.githubusercontent.com/2191036/80265973-61868b80-864e-11ea-9788-03b1261c072a.png)

Here's the line calling `SetMultiSamples`, only sets if the render window has been create:
https://github.com/CDAT/vcs/blob/a06bb9823aa6c96cc432bf47cf09072794d26786/vcs/VTKPlots.py#L1616-L1619

otherwise it's set when the render window is create:
https://github.com/CDAT/vcs/blob/a06bb9823aa6c96cc432bf47cf09072794d26786/vcs/VTKPlots.py#L623-L636

**Comment by jasonb5** [comment]
Installing the following creates a working environment:

```bash
conda create -n vcs-mesa-18.3.1_2 -c cdat/label/vtk_dev -c conda-forge/label/vtk_dev -c conda-forge vcs=8.2.2020.04.20.14.05.g2383318 "vtk=8.2.0" "mesalib=18.3.1" pillow
```
AA set to 0.
![vcs-aa-mesa_18 3 1-0-0](https://user-images.githubusercontent.com/2191036/80388395-b26cce80-885e-11ea-84ad-c8dbc3346bea.png)
AA defaulting to 8 for the baseline.
![vcs-aa-mesa_18 3 1-baseline-8](https://user-images.githubusercontent.com/2191036/80388403-b4369200-885e-11ea-9f34-dabd65bc620b.png)
Shows the difference.
![vcs-aa-mesa-baseline-aa0](https://user-images.githubusercontent.com/2191036/80388415-b6005580-885e-11ea-8150-fa87e6d5da86.png)

**Comment by jasonb5** [comment]
@muryanto1 Here's the condensed info, in the broken environment we cannot set antialiasing through the `SetMultiSamples` call but this does work in the working environment. Only difference is the VTK version. The `aa_diff.png` is blank in the broken environment and shows a difference in the working one.

Broken environment:
``` bash
conda create -n vcs-mesa-18.3.1-9.0.0 -c cdat/label/vtk_dev -c conda-forge/label/vtk_dev -c conda-forge vcs=8.2.2020.04.20.14.05.g2383318 "vtk=9.0.0.rc1" "mesalib=18.3.1" pillow
```

Working environment: 
```bash
conda create -n vcs-mesa-18.3.1-8.2.0 -c cdat/label/vtk_dev -c conda-forge/label/vtk_dev -c conda-forge vcs=8.2.2020.04.20.14.05.g2383318 "vtk=8.2.0" "mesalib=18.3.1" pillow
```

Scripts:
```python
import vtk
from PIL import ImageChops
from PIL import Image

colors = vtk.vtkNamedColors()

def draw_line(prefix, aa=None):
    renderer = vtk.vtkRenderer()

    window = vtk.vtkRenderWindow()
    window.SetSize(640, 480)
    window.AddRenderer(renderer)

    source = vtk.vtkLineSource()
    source.SetPoint1([0.0, 0.0, 0.0])
    source.SetPoint2([640.0, 480.0, 0.0])

    mapper = vtk.vtkPolyDataMapper()
    mapper.SetInputConnection(source.GetOutputPort())

    actor = vtk.vtkActor()
    actor.GetProperty().SetLineWidth(1.0)
    actor.GetProperty().SetColor(colors.GetColor3d('Black'))
    actor.SetMapper(mapper)

    renderer.AddActor(actor)
    renderer.SetBackground(colors.GetColor3d('White'))

    postfix = ''

    if aa is not None:
        window.SetMultiSamples(aa)

        postfix = f'_{aa}'

    value = window.GetMultiSamples()

    postfix = f'{postfix}_{value}'

    window.Render()

    image = vtk.vtkWindowToImageFilter()
    image.SetInput(window)
    image.SetInputBufferTypeToRGB()
    image.ReadFrontBufferOff()
    image.Update()

    filename = f'{prefix}{postfix}.png'

    writer = vtk.vtkPNGWriter()
    writer.SetFileName(filename)
    writer.SetInputConnection(image.GetOutputPort())
    writer.Write()

    return filename

base = draw_line('aa')
aa32 = draw_line('aa', 32)

img1 = Image.open(base)
img2 = Image.open(aa32)

diff = ImageChops.difference(img1, img2)

diff.save('aa_diff.png')
```

**Comment by jasonb5** [comment]
Here's the topic on VTK's discourse https://discourse.vtk.org/t/vtk-9-0-0rc3-python-cant-set-multisamples/3223. Will move to a bug once I clarify the usage.

**Comment by jasonb5** [comment]
Opened a bug on VTK's gitlab https://gitlab.kitware.com/vtk/vtk/-/issues/17891

**Comment by jasonb5** [comment]
When `setantialiasing` is called, vcs is calling `SetMultiSamples` on a `RenderWindow`.

We had three issues:
1. Plotting data using the default antialiasing value on linux and osx was causing thicker outlines as compared to a [reference image](https://cdat-vcs.readthedocs.io/en/latest/notebooks/boxfill_notebook.html)
2. Setting antialiasing in VTK 8.2.0 and 9.0.0 environments created different results (documented above) on linux system
3. Unittest was failing when checking the default antialiasing value, vcs sets it to 8 but the method returned 0, which is disabled.

Key points from the VTK issue:
- If `SetMultiSamples` is called with 32 and your driver/hardware only supports 8 AA (antialiasing) then the VTK library will set the value to 8; the maximum supported value. A subsequent call to `GetMultiSamples` would return 8 and only values less than or equal to 8 are valid.
- VTK has changed how they determine the maximum AA value, in 8.2.0 they would call `glXChooseFBConfig` with a large MultiSample value and continue calling this with smaller values until the call succeeded, in 9.0.0 they now call `glGetIntegerv(GL_MAX_SAMPLES, &msamples)`. I'm not sure of the full effects of this since it's dependent on the OpenGL implementation and driver/hardware which can change.

Issue resolutions:
1. I'm rechecking this.
2. Aftter some debugging with a VTK developer we determined this was caused due to competing OpenGL libraries; 8.2.0 was using one and 9.0.0 was using the other. These two OpenGL libraries returned different results for the maximum AA value, causing the difference in outputs.
3. We need to update the unit test, which is being run on headless machine and using software OpenGL. The unittest calls `open` which calls `SetMultiSamples` with 8 and we check that the value is 8. This doesn't take into consideration what the maximum possible value is e.g. if the OpenGL implementation doesn't have AA enabled it's possible for it to return 0 or if it's supported value is less than 8. We could possibly use pyopengl or pyglet to get a valid value for the setup.

**Comment by jasonb5** [comment]
Forgot to mention according to my conversation with the VTK developer antialiasing does not work on headless machines, the behavior should be the same as setting antialiasing to 0 on machines with graphics hardware.

**Comment by jasonb5** [comment]
Fixes the unittest in e78fed342e8f5648a2e6ed403f8f69c64af2f217.
